### ISSUE TITLE
Issue with Jsoncons library

### ISSUE THREAD

**Issue by BojanKV** [rollbear/trompeloeil#167]
Hi, 

There is a compilation issue with Jsoncons and trompeloeil library which I'm unable to resolve.

Jsoncons is a C++ header-only library and we use it extensively in our Eclipse KUKSA project.
Tested with older local version and with latest version from Jsoncons Gihtub [project](https://github.com/danielaparker/jsoncons).

We are currently in a process of improving our testing infrastructure and would like to use trompeloeil as a mock utility together with existing Boost::test library.

Demo files and code is located local Github branch [here](https://github.com/BojanKV/kuksa.invehicle/tree/9ece03df32f9e2adadf6ea19e7e89c15a6dbff65/w3c-visserver-api/test/issue).

Compiled with command:
`g++ -std=c++14 -I. -c demo.cpp -o demo`

Compilation log: [build.log](https://github.com/rollbear/trompeloeil/files/3797633/build.log)

Thanks in advance,
Bojan

**Comment by rollbear** [comment]
There's an unfortunate mismatch between the Trompeloeil metaprogramming to figure out how to print an unknown type, and Jsoncons's metaprogramming to provide nice and meaningful compilation error messages when you try to instantiate with the wrong types. I'm leaning towards trying to make Jsoncons's handling a bit more SFINAE friendly, but I'm not at all familiar with the code, and I'm not entirely certain that it's the right way to go.

Just thought I'd let you know I haven't forgotten.

**Comment by rollbear** [comment]
OK, I have a possible solution, but I'm not happy with it. It's a change to Jsoncons, so it should be reported to them.

```diff
diff --git a/include/jsoncons/basic_json.hpp b/include/jsoncons/basic_json.hpp
index dd41af07..1bdd1740 100644
--- a/include/jsoncons/basic_json.hpp
+++ b/include/jsoncons/basic_json.hpp
@@ -2942,6 +2942,9 @@ public:
     {
     }
 
+    basic_json(std::nullptr_t) = delete;
+    basic_json(std::nullptr_t, const Allocator&) = delete;
+    basic_json(std::nullptr_t, semantic_tag) = delete;
     template <class T>
     basic_json(const T& val, const Allocator& allocator)
         : var_(json_type_traits<basic_json,T>::to_json(val,allocator).var_)
```

I tried to use `enable_if`, to SFINAE-out the templated constructor, but that doesn't work, because it just removes it from the overload set, and selects the `const char*` version of constructor instead. This gives a runtime error, which is way nastier.

I believe better solutions exists, but they will boil down to attempting construction of a `json` object from `nullptr` resulting in the error that no constructor was found. The original implementation does find a constructor, but you get a static_assert triggered when calling it is attempted. That's the error you found.

**Comment by rollbear** [comment]
Can you check if you still have problems with the just released v36? The fix for issue #173 probably takes care of this one too.

**Comment by rollbear** [comment]
Rechecked, and this now works with latest on `jsoncons` and tagged v39 of Trompeloeil.
